### PR TITLE
[dag] Fix caching for `apply_recursive`

### LINKED ISSUE #28163: [DAG] The complexity of basic Ray DAG operations are exponential
### What happened + What you expected to happen

Due to incorrect implementation, when calling `node.apply_recursive` (which is essential and basic to most DAG operations), the operation time is exponential. So it looks like running forever for even slightly bigger DAGs.

To understand what is happening, the algorithm in current code would traversing every possible paths starting between the output and inputs nodes in the DAG (the caching still works, but it is hit exponentially).

The solution is simple:

```python

    def apply_recursive(self, fn: "Callable[[DAGNode], T]", _cache: Dict = None) -> T:
        """Apply callable on each node in this DAG in a bottom-up tree walk.

        Args:
            fn: Callable that will be applied once to each node in the
                DAG. It will be applied recursively bottom-up, so nodes can
                assume the fn has been applied to their args already.

        Returns:
            Return type of the fn after application to the tree.
        """

        if _cache is None:
            _cache = {}

        if type(self).__name__ == "InputNode":
            if "input" in _cache:
                raise AssertionError("Each DAG should only have one unique InputNode.")
        if self._stable_uuid in _cache:
            return _cache[self._stable_uuid]

        replace_table = {}
        # CloudPickler scanner object for current layer of DAGNode. Same
        # scanner should be use for a full find & replace cycle.
        scanner = _PyObjScanner()

        # Find all first-level nested DAGNode children in args.
        # Update replacement table and execute the replace.
        for node in scanner.find_nodes(
            [
                self._bound_args,
                self._bound_kwargs,
                self._bound_other_args_to_resolve,
                self._bound_upstream_nodes,
            ]
        ):
            if node not in replace_table:
                replace_table[node] = node.apply_recursive(fn, _cache)
        (
            new_args,
            new_kwargs,
            new_other_args_to_resolve,
            new_upstream_nodes,
        ) = scanner.replace_nodes(replace_table)

        # Return updated copy of self.
        new_node = self._copy(
            new_args,
            new_kwargs,
            self.get_options(),
            new_other_args_to_resolve,
            new_upstream_nodes,
        )

        output = fn(new_node)
        if type(self).__name__ == "InputNode":
            _cache["input"] = output
        else:
            _cache[self._stable_uuid] = output
        return output

```

### Versions / Dependencies

latest upstream

### Reproduction script

```python
import ray
from ray.dag import DAGNode


origin_apply_recursive = DAGNode.apply_recursive

counter = 0


def _apply_recursive(self, fn):
    global counter
    counter += 1
    return origin_apply_recursive(self, fn)


DAGNode.apply_recursive = _apply_recursive


if __name__ == '__main__':
    @ray.remote
    def join(a, b):
        return None

    a, b = None, None
    n = 10
    for _ in range(n):
        a, b = join.bind(a, b), join.bind(a, b)

    a.apply_recursive(lambda x: x)
    assert counter == 2 ** n - 1
```

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

### PR DESCRIPTION
**Pull request by michaelhly**
## Why are these changes needed?
Caching was broken when applying a callable on a DAGNode, leading to exponential runtime. 

## Related issue number

Closes #28163

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/dag/dag_node.py
Comment (michaelhly) [inline python/ray/dag/dag_node.py:257]: A new cache was created on each `apply_recursive`.

**Comment by SebastianMorawiec** [comment]
Hello @michaelhly - This PR would resolve high-impact issue we are currently having with Ray. I'd really appreciate if this could be merged until the next release. If you do not have time for that, I could take over due diligence for that PR.

**Comment by michaelhly** [comment]
@SebastianMorawiec I'm just some random contributor, you will have to ping someone on the team to get this resolved.

**Comment by SebastianMorawiec** [comment]
@jjyao could I bring some attention to this? I'd really appreciate this PR being merged, its seems simple enough and fixes high priority issue for us.

**Review by jjyao — APPROVED** [review]

**Comment by jjyao** [comment]
@michaelhly Thanks for the contribution and sorry for missing this PR. Could you mark it as ready for review and rebase with master to fix conflicts. After that, I'll merge the PR.

**Review by jjyao — APPROVED** [review]
I'll merge after all tests pass.

**Pull request merged**
